File: typescript/src/Asset.ts
export class Asset {
    private readonly _description: string;
    private readonly _date: Date;
    private _value: Value;

    public constructor(description: string, date: Date, value: Value) {
        this._description = description;
        this._date = date;
        this._value = value;
    }

    public getDescription(): string {
        return this._description;
    }

    public getDate(): Date {
        return this._date;
    }

    public getValue(): Value {
        return this._value;
    }

    setValue(value: Value): void {
        this._value = value;
    }
}

export abstract class Value {
    protected readonly _value: number;

    protected constructor(value: number) {
        this._value = value;
    }

    public get(): number {
        return this._value;
    }

    public ToString(): string {
        return this._value.toString();
    }
}

export class MeasurableValue extends Value {
    constructor(value: number) {
        super(value);
    }
}

export class PricelessValue extends Value {

    constructor() {
        super(Number.POSITIVE_INFINITY);

    }
}

export class NoValue extends Value {
    constructor() {
        super(0);
    }
}

File: c_sharp/Portfolio/Asset.cs
using System.Globalization;

namespace Portfolio;

public class Asset
{
    private readonly DateTime _date;

    public Asset(string description, DateTime date, Value value)
    {
        Description = description;
        _date = date;
        Value = value;
    }

    public string Description { get; }

    public DateTime Date => _date.Date;

    public Value Value { get; set; }
}

public abstract class Value
{
    protected readonly float _value;

    protected Value(float value)
    {
        _value = value;
    }

    public float Get()
    {
        return _value;
    }
}

public class MeasurableValue: Value
{
    public MeasurableValue(float value) : base(value)
    { }
    
    public override string ToString()
    {
        return _value.ToString(CultureInfo.CurrentCulture);
    }
}

public class PricelessValue : Value 
{
    public PricelessValue() : base(float.PositiveInfinity)
    { }
}

public class NoValue : Value 
{
    public NoValue() : base(0)
    { }
}

